### ISSUE TITLE
Freedom may prioritize secondary_targets over expansion

### ISSUE THREAD

**Issue by SurDno** [Decane/SRP#62]
I slugged the whole Duty Base and then started waiting for a Freedom squad to appear. I was literally waiting for an hour (I left the game open and got away), though nothing happened. All of newly spawned squads headed to **Barrier** and captured some other points on Army Warehouses instead. 

[Here is the save](https://yadi.sk/d/4-yWxOIKqv6g7g) where you can see this happening.

_P.S. — I do not use the latest SRP branch, so this may actually be a duplicate of #23, please, check it first. I've also already captured this base with bandits with a modified faction_bandit.ltx, that may be a reason for this bug too. No fix required then, it's my fault._

**Comment by SurDno** [comment]
Ok, this really got out of control. [Here is the save](https://yadi.sk/d/q_JZYdM4CQqPDg).

Freedom is literally capturing everything but Duty Base. But loners are trying to capture Duty Base exits (even while their own base is populated by bandits). 
![image](https://user-images.githubusercontent.com/35204405/49021976-ded17680-f1a4-11e8-9c96-333f215694b4.png)

**Comment by Decane** [comment]
@SurDno You have highlighted two problems:

1. Precondition targets are not given special special treatment over secondary targets (as when Freedom goes to the Barrier instead of the Duty base).

2. Precondition targets are not given special special treatment over power targets (as when Loners try to capture resource/territory smart terrains instead of their own base).

I've fixed both (and several other prioritization problems) in the current WIP branch: https://github.com/Decane/SRP/commits/WIP. I'll merge to the master branch once I've confirmed I haven't regressed anything.

**Comment by Decane** [comment]
The game's squad assignment system has three major components:

1. function `faction_brain_human:calculate_expansion_targets()`

Populates a table named "target_smart_value" for a given faction by adding smart terrains to it in the following fashion:

- Secondary targets (specified under `[secondary_targets]` in faction_*.ltx) are added without tasking the player to eliminate/capture them.

- Territory/resource targets available according to `sim_board.is_point_avail()` are added if (i) owned by the faction or (ii) the faction's squads' strength relative to their main opponent's is below the `precondition_power` value specificed in faction_*.ltx.

  If the player is a member of the faction and the point is captured by a faction hostile to the player's faction, the player is tasked with either:

  - (a) capturing the point if allies close to it from the player's faction are also vying for it, or
  - (b) eliminating the point if no allies are near it _and_ the player is within 300m of the point on the same level.

- Precondition targets (specified per expansion via parameter `precondition_target` in faction_*.ltx) for the faction's current expansion and below are added with a flag that tasks the player to capture the target if it's empty or captured by hostiles.

- Eliminate targets (specified per expansion via parameter `precondition_target_eliminate` in faction_*.ltx) for the faction's current expansion and below are added if captured by hostiles. They are added with a flag that tasks the player to eliminate the target if captured by hostiles.

So there are 4 'categories' of faction war target. Each of them is added with a `point_capacity` (how many squads they support), which equals:

  - (a) the point's actual maximum squad capacity as specified in all.spawn if the point is captured by a faction other than the player's;
  - (b) the point's actual maximum capacity minus one if captured by the player's faction or no-one at all and the point is an important one (`important_point = true` in all.spawn) and the point is a potential respawn point for any faction whatsoever;
  - (b) 1 if captured by the player's faction or no-one at all and the point is **not** an important one.

2. function `se_sim_faction:calculate_squad_tasks()`

Populates a table named "squad_target_cache" which maps squads to a target point added to table "target_smart_value" in the previous step. Any squads with a scripted target immediately get mapped to that. Any squads without a scripted target are first treated as potential assignees to any points that they are on the same 'level cluster' with.

Then, the entire sub-set of "target_smart_value" points with at least one available squad -- one squad may count as 'available' to several points at this stage -- is sorted in ascending order of distance between point and squad closest to that point. The first point after this sorting is usually one with 0m between it and the squad closest to it - that is, one whose closest squad is already on it.

The closest squad is then assigned to that point if the point's squad capacity allows it, and the squad is made unavailable to all other points in table "target_smart_value". If the point's squad capacity was saturated thereby, it's removed from the pool of points to assign a squad to. Then, the set of points with at least one available squad is again sorted by distance between its points and their closest squad and the next squad-point pair with the least distance between them gets mapped in table "squad_target_cache".

This process continues until there are no more available squads left to assign to a point or until there are no more target points needing a squad.

3. function `sim_squad_generic:get_next_action()`

This function ensures any squads that didn't get a target in the previous step stay put, and either sends or doesn't send squads that did get a target to their target according to various conditions, including the 'desirability' of the intermediate camps between the squad and its target. I could write a thesis on how that happens...

**Comment by SurDno** [comment]
How special treatment should be to be considered special special? :) There are also two (b) behaviors where you explain points’ capacity.

Why do (b) and (c) behaviors even exist? Are they the reason for #70?

**Comment by SurDno** [comment]
How special treatment should be to be considered special special? :) There are also two (b) behaviors where you explain points’ capacity.

Why do (b) and (c) behaviors even exist? Are they the reason for #70?

How does “level cluster” mechanic works? Why does the game sometimes send a squad from another end of the map?

When I modified config files for factions myself, I noticed while making bandits capture Agroprom that they do not even try to capture points other than stated in configs (i.e. they head directly to the Duty Base without capturing other important and resource points). Why does that happen?

Why does the faction not try to capture other important points after it captured the enemy base?

And last but not least, is it possible (without engine edits and significant code changes) to unlock the Faction War to make all factions capture all the points all around the Zone? Similar happened with Freedom on my current broken save (I showed it to you, the one where the Flea Market is captured by two factions) and I really want to see a warfare with all the factions doing it. :)

And last: is it possible to make factions have more than 4 resource tiers without engine edits?

**Comment by Decane** [comment]
I forgot to mention that the faction prioritization behavior described in my previous post is accurate for the vanilla game and the current SRP master branch, but not entirely accurate for the aforementioned WIP branch, since there I have introduced a more sophisticated sorting method for (2) which divides the four types of faction target into priority 'tiers' and sorts the targets one tier at a time (by distance between target and closest squad, as before):

Eliminate targets ~ Precondition targets > Power targets > Secondary targets

> There are also two (b) behaviors where you explain points’ capacity.

D'oh! Fixed, thanks. :)

> Why do (b) and (c) behaviors even exist? Are they the reason for #70?

Suppose a point is occupied by 1 Clear Sky squad and is about to be occupied by a further 2 Clear Sky squads, who are just about to enter it.

(c) probably exists because demoting the point's squad capacity to 1 in Clear Sky's eyes now that 1 squad is already on it permits the 2 'extra' Clear Sky squads heading to that point to be re-assigned to capture other points, some of which may be precondition targets for further faction expansion. Whereas if the point's capacity were to stay at 3 even after being captured, all 3 squads would always keep being re-assigned to that point since they would all be 0m away from it. This would possibly leave insufficiently many squads available to capture other targets necessary for advancing the faction war.

Now suppose the same point is occupied by no-one. Perhaps GSC thought it makes little sense for Clear Sky to send 3 squads to the point if capturing it will be met with no resistance. So, its point capacity gets demoted to 1 for Clear Sky to ensure they only send 1 squad to capture it.

(b) probably exists for the same reasons as (c), but with the difference that if the target point is ultra-important, Clear Sky probably wants to occupy it fully even if it's currently empty or already captured by Clear Sky, in order to secure it in the event of attacks by hostiles. The reason why its point capacity is set to the maximum capacity **minus one** is probably because if it's a respawn sector, it can only be used to respawn squads if it's not completely full, i.e. if the number of squads occupying it is at most point_capacity - 1.

I'm not aware of any changes in the current SRP master branch that would explain https://github.com/Decane/SRP/issues/70. Are you sure vanilla never exhibits the same behavior?

> How does “level cluster” mechanic works?
```
sim_level_cluster = {

	marsh                = 1,
	escape               = 2,
	garbage              = 2,
	darkvalley           = 2,
	agroprom             = 2,
	agroprom_underground = 3,
	yantar               = 2,
	red_forest           = 2,
	military             = 2,
	limansk              = 4,
	hospital             = 5,
	stancia_2            = 6
}
```
It's simple: in step (2) of my previous post, a squad will only be marked as suitable for assignment to one of the 4 types of target point in table "target_smart_value" if they are on a map with the same numeric value in the table above as the map on which the target point resides.

See how "marsh" has value 1? That's the reason why you don't see Clear Sky attempting to capture points outside of the Swamps even if they have a clear path to such points.

> Why does the game sometimes send a squad from another end of the map?

Remember those 4 'types' of faction target outlined in my previous post? In vanilla and the current SRP master branch, they are all prioritized _only_ by distance to their closest squad. So squads will refuse to leave already captured targets of even the 'power target' type (of which there are tons on almost every map) despite a neighboring point being essential for advancement in the faction war. So the faction has no choice but to assign a further-away squad to the target. (The current WIP branch "fixes" this by making factions always prioritize precondition targets over power targets.)

> When I modified config files for factions myself, I noticed while making bandits capture Agroprom that they do not even try to capture points other than stated in configs (i.e. they head directly to the Duty Base without capturing other important and resource points). Why does that happen?

Not sure without playing such a save myself and enabling some debug logging to show what's going on under the covers.

> Why does the faction not try to capture other important points after it captured the enemy base?

Because the expansion subsequent to the one that required the base capture does not result in any new targets being added to the faction's "target_smart_value" table. :)

> And last but not least, is it possible (without engine edits and significant code changes) to unlock the Faction War to make all factions capture all the points all around the Zone?

Yes, try setting `precondition_power = 1` for the last faction expansion in faction_*.ltx. That will make the faction stop trying to capture points only once _every single squad_ of their main opposing faction has been annihilated.

> And last: is it possible to make factions have more than 4 resource tiers without engine edits?

Yes, trivial even. Just add more in configs\gameplay\character_desc_*.xml and reference them in squad_descr_*.ltx or `[spawn_by_resource]` in faction_*.ltx.
